### PR TITLE
Solve Chapter1

### PR DESCRIPTION
**Pull request by dit7ya**
### Solutions for Chapter 1

cc @vrom911 @chshersh

### CODE REVIEW

**Comment by vrom911** [comment]
Hey @dit7ya . Would you like us to review your solutions? It is totally okay in both ways, I just noticed that you mentioned us in the PR.  
Also, feel free to ask any questions, we are here to help 🙂

**Comment by dit7ya** [comment]
@vrom911 First of all, thank you very much for creating this amazing self-contained learning experience. I will be super happy to have my solutions reviewed. Also, I have the following question

- What is the design rationale in Haskell for allowing only one output for functions (while multiple input arguments are supported)?

File: src/Chapter1.hs [hunk 1]
```diff
@@ -556,7 +579,10 @@ True
 >>> isVowel 'x'
 False
 -}
-isVowel c = error "isVowel: not implemented!"
+isVowel :: Char -> Bool
+isVowel c
+  | elem c "aeiou" = True
```
Comment (vrom911) [inline src/Chapter1.hs:584]: Nice one! 👍🏼 
As `elem` returns `Bool` itself, you can just return the result of this function:

```haskell
isVowel c = elem c "aeiou"
```

One note in here, that sometimes, `elem` could be slower than the explicit pattern matching. I remember there were some benchmarks on one particular case, that showed how moving to pattern matching on each case separately drastically decrease time 🐎

File: src/Chapter1.hs [hunk 2]
```diff
@@ -620,7 +646,11 @@ Try to introduce variables in this task (either with let-in or where) to avoid
 specifying complex expressions.
 -}
 
-sumLast2 n = error "sumLast2: Not implemented!"
+sumLast2 :: Int -> Int
+sumLast2 n = theLastDigit + secondLastDigit
+  where
+    theLastDigit = abs n `mod` 10
+    secondLastDigit = (abs n `mod` 100 - theLastDigit) `div` 10
```
Comment (vrom911) [inline src/Chapter1.hs:653]: That is a wonderful solution! 👏🏼  You correctly noticed that it is the `div` and `mod`, cool 😎 

One hint to make your solution even shorter: the standard library has the [`divMod`](https://hackage.haskell.org/package/base-4.14.0.0/docs/Prelude.html#v:divMod) function, that is actually combines inside both `div` and `mod`. And this is exactly what you use!.

So you could write it this way:

```haskell
(theLastDigit, secondLastDigit) = x `divMod` 10
```

You can see how we could pattern match on the pair 🙂

File: src/Chapter1.hs [hunk 3]
```diff
@@ -641,7 +671,10 @@ You need to use recursion in this task. Feel free to return to it later, if you
 aren't ready for this boss yet!
 -}
 
-firstDigit n = error "firstDigit: Not implemented!"
+firstDigit :: Int -> Int
+firstDigit n
+  | abs n < 10 = abs n
+  | otherwise = firstDigit ((abs n - abs n `mod` 10) `div` 10)
```
Comment (vrom911) [inline src/Chapter1.hs:677]: No need to subtract `abs n nod 10` as it won't affect the `div` result 🙂 

On the other note, you reuse `abs n` a lot in this function. That usually means that you can move it into a separate variable in the `where` clause, and use everywhere 🙂

**Review by vrom911 — COMMENTED** [review]
Nice work on Chapter Two 🥇 

Regarding your question, functions in Haskell always have just one input and one output.
You partially saw that when we were talking about partial application.

So, in reality, function like

```haskell
foo :: a -> b -> c -> d
```

You may say that it has three input arguments and one output.
But actually, it is not quite like that for the compiler. It is always 1 input - 1 output. It only the matter of the parenthesis. In Haskell, the arrow (`->`) that represents the function if left-associative. So, if you fancy adding the parenthesis to the above `foo` function, it would look like this:

```haskell
foo :: a -> (b -> (c -> d))
```

And it is actually the way  __every__ function in Haskell works 🙂 

Hope this is what you wanted to clarify. Let me know if you have any questions 🙂

**Pull request merged**